File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
res=987654321
v=[0 for _ in range(51)]
def chk(s1,s2):
    c=0
    if len(s1)!=len(s2):
        return False
    for i in range(len(s1)):
        if s1[i]!=s2[i]:
            c+=1
    if c>1:
        return False
    return True
def go(s,t,cnt,w):
    global res
    if s==t:
        res=min(res,cnt)
        return
    for i in range(len(w)):
        if not v[i] and chk(s,w[i]):
            v[i]=1
            go(w[i],t,cnt+1,w)
            v[i]=0

def solution(begin, target, words):
    answer = 0
    global res
    if not target in words:
        return 0
    go(begin,target,0,words)
    return res

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
class Solution {
    static int []visited=new int[51];
    static String t;
    static int res=987654321;
    static String[] w;
    public boolean chk(String s,String s2)
    {
        int cnt=0;
        for (int i=0;i<s.length();i++)
        {
            if (s.charAt(i)!=s2.charAt(i))
                cnt+=1;
            if (cnt>1)
            return false;
        }
        return true;
    }
    public void go(String s,int cnt)
    {
        if (s.equals(t))
        {
            res=Math.min(res,cnt);
            return;
        }
        for (int i=0;i<w.length;i++)
        {
            if (visited[i]==0&&chk(s,w[i]))
            {
                visited[i]=1;
                go(w[i],cnt+1);
                visited[i]=0;
            }
        }
        
    }
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        t=target;
        w=words;
        go(begin,0);
        if (res==987654321)
            res=0;
        return res;
    }
}